miniapps/esl-alirev.man
.TH "esl-alirev" 1 "@RELEASEDATE@" "@PACKAGE@ @RELEASE@" "@PACKAGE@ Manual"

.SH NAME
.TP
esl-alirev - reverse complement a multiple alignment 


.SH SYNOPSIS

.B esl-alirev
.I [options]
.I msafile


.SH DESCRIPTION

.pp
.B esl-alirev
reads the multiple alignment in
.I msafile
and outputs its reverse complement to 
.I stdout.

.pp
An example of where you might need to do this is when you've
downloaded a chunk of multiway genomic alignment from one of the
genome browsers, but your RNA of interest is on the opposite strand.

.pp
Any per-column and per-residue annotation lines are reversed as well,
including Stockholm format and old SELEX format annotations.
Annotations that Easel recognizes as secondary structure annotation (a
consensus structure line, individual secondary structure lines) will
be "reverse complemented" to preserve proper bracketing orders: for
example, ...<<<...>>> is reverse complemented to <<<...>>>..., not
simply reversed to >>>...<<<..., which would be wrong.

.pp
If
.I msafile 
is - (a single dash), alignment input is read from 
.I stdin.

.pp 
The
.I msafile
may be in any of several formats.
The format is autodetected by default.
See the
.B --informat
option to assert an input format.

.pp 
By default the output alignment is written in the same format as the
input alignment. See the
.B --outformat
option to use a different output format.

.pp
Because the alignment is parsed into Easel's digital internal
representation, the output alignment may differ in certain details
from the original alignment; these details should be inconsequential
but may catch your eye. One is that if you have a reference annotation
line, Easel's output will put consensus residues in upper case,
nonconsensus (inserted) residues in lower case. Another is that the
headers for some formats, such as Clustal format, are written with an
arbitrary version number -- so you may find yourself revcomping an
alignment in "MUSCLE (3.7) multiple sequence alignment" format and it
could come out claiming to be a "CLUSTAL 2.1 multiple sequence
alignment", just because Easel writes all of its Clustal format
alignment files with that header.

.pp
The
.I msafile
must contain nucleic acid sequences (DNA or RNA).
The alphabet will be autodetected by default.
See the
.B --dna
or
.B --rna 
options to assert an alphabet.



.SH OPTIONS

.TP
.B -h 
Print brief help;  includes version number and summary of
all options, including expert options.

.TP
.BI --informat " <s>"
Specify that the input alignment is in 
.I <s>
format. Valid format strings include 
"stockholm", "pfam", "a2m", "phylip", "phylips", 
"psiblast", "selex", "afa", "clustal", and "clustallike".

.TP
.BI --outformat " <s>"
Write the output alignment in 
.I <s>
format. Valid format strings include 
"stockholm", "pfam", "a2m", "phylip", "phylips", 
"psiblast", "selex", "afa", "clustal", and "clustallike".

.TP 
.B --dna
Assert that the 
.I msafile 
contains DNA sequences. 

.TP 
.B --rna
Assert that the 
.I msafile 
contains RNA sequences. 


.SH AUTHOR

Easel and its documentation are @EASEL_COPYRIGHT@.
@EASEL_LICENSE@.
See COPYING in the source code distribution for more details.
The Easel home page is: @EASEL_URL@


miniapps/esl-alirev.c


#include "easel.h"
#include "esl_alphabet.h"
#include "esl_getopts.h"
#include "esl_msa.h"
#include "esl_msafile.h"
#include "esl_vectorops.h"

#include <stdio.h>

static ESL_OPTIONS options[] = {
  /* name             type          default  env  range toggles reqs incomp  help                                       docgroup*/
  { "-h",          eslARG_NONE,       FALSE,  NULL, NULL,  NULL,  NULL, NULL, "show brief help on version and usage",        0 },
  { "--informat",  eslARG_STRING,      NULL,  NULL, NULL,  NULL,  NULL, NULL, "specify the input MSA file is in format <s>", 0 }, 
  { "--outformat", eslARG_STRING,      NULL,  NULL, NULL,  NULL,  NULL, NULL, "write the output MSA in format <s>",          0 }, 
  { "--dna",       eslARG_NONE,       FALSE,  NULL, NULL,  NULL,  NULL, NULL, "use DNA alphabet",                            0 },
  { "--rna",       eslARG_NONE,       FALSE,  NULL, NULL,  NULL,  NULL, NULL, "use RNA alphabet",                            0 },
  {  0, 0, 0, 0, 0, 0, 0, 0, 0, 0 },
};
static char usage[]  = "[-options] <msafile>";
static char banner[] = "reverse complement multiple sequence alignment(s)";

int
main(int argc, char **argv)
{
  ESL_GETOPTS  *go      = esl_getopts_CreateDefaultApp(options, 1, argc, argv, banner, usage);
  char         *msafile = esl_opt_GetArg(go, 1);
  int           infmt   = eslMSAFILE_UNKNOWN;
  int           outfmt  = eslMSAFILE_UNKNOWN;
  ESL_ALPHABET *abc     = NULL;
  ESLX_MSAFILE *afp     = NULL;
  ESL_MSA      *msa     = NULL;
  int           nali    = 0;
  int           status;

  /* Alphabet specification from cmdline? */
  if      (esl_opt_GetBoolean(go, "--rna"))   abc = esl_alphabet_Create(eslRNA);
  else if (esl_opt_GetBoolean(go, "--dna"))   abc = esl_alphabet_Create(eslDNA);

  /* MSA input file format from cmdline? */
  if (esl_opt_IsOn(go, "--informat") &&
      (infmt = eslx_msafile_EncodeFormat(esl_opt_GetString(go, "--informat"))) == eslMSAFILE_UNKNOWN)
    esl_fatal("Your --informat, %s, is not a recognized multiple alignment file format",
	      esl_opt_GetString(go, "--informat"));

  /* Open in digital mode. Autoguess alphabet, format if we haven't set them already. */
  if (( status = eslx_msafile_Open(&abc, msafile, NULL, infmt, NULL, &afp)) != eslOK)
    eslx_msafile_OpenFailure(afp, status);
  
  /* Reverse complementation only makes sense for alphabets that have abc->complement set */
  if (! abc->complement)
    esl_fatal("File %s appears to use the %s alphabet.\nThat alphabet cannot be reverse complemented.\n",
	      msafile, esl_abc_DecodeType(abc->type));

  /* Set the output format, if requested. 
   * By default, write in the same format we read in. 
   * Remember, it's afp->format that gets set; infmt was only a hint.
   */
  if ( esl_opt_IsOn(go, "--outformat") )
    {
      outfmt = eslx_msafile_EncodeFormat(esl_opt_GetString(go, "--outformat"));
      if (outfmt == eslMSAFILE_UNKNOWN)
	esl_fatal("Your --outformat, %s, is not a recognized multiple alignment file format.\n",
		  esl_opt_GetString(go, "--outformat"));
    }
  else outfmt = afp->format;

  /* Here we go. */
  while ((status = eslx_msafile_Read(afp, &msa)) == eslOK)
    {	
      nali++;

      status = esl_msa_ReverseComplement(msa);

      eslx_msafile_Write(stdout, msa, outfmt);

      esl_msa_Destroy(msa);
    }
   if (nali   == 0)      esl_fatal("No alignments found in input file %s\n", msafile);
   if (status != eslEOF) eslx_msafile_ReadFailure(afp, status);

   eslx_msafile_Close(afp);
   esl_alphabet_Destroy(abc);
   esl_getopts_Destroy(go);
   return 0;
}
